### PR TITLE
Add an assertion that demangle "trees" are DAGs (NFC)

### PR DESCRIPTION
**Pull request by adrian-prantl**
rdar://73176262

### CODE REVIEW

**Comment by adrian-prantl** [comment]
@swift-ci test

File: lldb/source/Plugins/TypeSystem/Swift/TypeSystemSwiftTypeRef.cpp [hunk 1]
```diff
@@ -1611,9 +1611,37 @@ bool Equivalent(llvm::Optional<T> l, T r) {
...
+static bool ContainsCycles(swift::Demangle::NodePointer node) {
+  DAGCheckSet visited;
+  return ContainsCyclesImpl(visited, node);
+}
+#endif
+
 CompilerType
 TypeSystemSwiftTypeRef::RemangleAsType(swift::Demangle::Demangler &dem,
-                                       swift::Demangle::NodePointer node) {
+                                       swift::Demangle::NodePointer node) __attribute__((optnone)){
```
Comment (kastiglione) [inline lldb/source/Plugins/TypeSystem/Swift/TypeSystemSwiftTypeRef.cpp:1644]: is `optnone` to be committed?

**Review by kastiglione — APPROVED** [review]

File: lldb/source/Plugins/TypeSystem/Swift/TypeSystemSwiftTypeRef.cpp [hunk 2]
```diff
@@ -1611,9 +1611,37 @@ bool Equivalent(llvm::Optional<T> l, T r) {
   return IMPL();
 #endif
 
+
+#ifndef NDEBUG
+using DAGCheckSet = llvm::SmallPtrSet<swift::Demangle::NodePointer, 8>;
+static bool ContainsCyclesImpl(DAGCheckSet &visited,
```
Comment (kastiglione) [inline lldb/source/Plugins/TypeSystem/Swift/TypeSystemSwiftTypeRef.cpp:1617]: The cycle  check may be a good addition to `swift::Demangle` itself. I'm assuming lldb isn't the only consumer that would want to ensure cycles don't accidentally occur.

File: lldb/source/Plugins/TypeSystem/Swift/TypeSystemSwiftTypeRef.cpp (hunk 1 above)
Comment (adrian-prantl) [reply lldb/source/Plugins/TypeSystem/Swift/TypeSystemSwiftTypeRef.cpp:1644]: 🙄

**Pull request closed without merging**